upf/pfcp.c

```diff
@@ -7753,13 +7754,126 @@ encode_group (const struct pfcp_ie_def *def, struct pfcp_group *grp,
   return r;
 }
 
+static inline int
+pfcp_is_session_msg (u16 type)
+{
+  switch (type)
+    {
+    case PFCP_HEARTBEAT_REQUEST:
+    case PFCP_HEARTBEAT_RESPONSE:
+    case PFCP_PFD_MANAGEMENT_REQUEST:
+    case PFCP_PFD_MANAGEMENT_RESPONSE:
+    case PFCP_ASSOCIATION_SETUP_REQUEST:
+    case PFCP_ASSOCIATION_SETUP_RESPONSE:
+    case PFCP_ASSOCIATION_UPDATE_REQUEST:
+    case PFCP_ASSOCIATION_UPDATE_RESPONSE:
+    case PFCP_ASSOCIATION_RELEASE_REQUEST:
+    case PFCP_ASSOCIATION_RELEASE_RESPONSE:
+    case PFCP_VERSION_NOT_SUPPORTED_RESPONSE:
+    case PFCP_NODE_REPORT_REQUEST:
+    case PFCP_NODE_REPORT_RESPONSE:
+      return 0;
+
+    case PFCP_SESSION_SET_DELETION_REQUEST:
+    case PFCP_SESSION_SET_DELETION_RESPONSE:
+    case PFCP_SESSION_ESTABLISHMENT_REQUEST:
+    case PFCP_SESSION_ESTABLISHMENT_RESPONSE:
+    case PFCP_SESSION_MODIFICATION_REQUEST:
+    case PFCP_SESSION_MODIFICATION_RESPONSE:
+    case PFCP_SESSION_DELETION_REQUEST:
+    case PFCP_SESSION_DELETION_RESPONSE:
+    case PFCP_SESSION_REPORT_REQUEST:
+    case PFCP_SESSION_REPORT_RESPONSE:
+      return 1;
+
+    default:
+      return -1;
+    }
+}
+
 int
-pfcp_encode_msg (u16 type, struct pfcp_group *grp, u8 ** vec)
+pfcp_decode_msg (u8 * p, int len,
+		 pfcp_decoded_msg_t * dmsg, pfcp_offending_ie_t ** err)
 {
-  assert (type < ARRAY_LEN (msg_specs));
-  assert (msg_specs[type].size == 0 || msg_specs[type].group != NULL);
+  u16 type, msg_len;
+  int is_session_msg;
+  uword hdr_len;
+
+  clib_memset (dmsg, 0, sizeof (*dmsg));
+
+  if (!len)
+    return -1;			/* empty message, can't generate any response */
+
+  type = pfcp_msg_type (p);
+  is_session_msg = pfcp_is_session_msg (type);
+  hdr_len = is_session_msg ? SESSION_MSG_HDR_LEN : NODE_MSG_HDR_LEN;
+
+  if (len < hdr_len)
+    return -1;			/* incomplete header, can't generate any response */
+
+  if (type >= ARRAY_LEN (msg_specs) ||
+      !(msg_specs[type].size == 0 || msg_specs[type].group != NULL))
+    return -1;			/* bad type, can't generate any response */
+
+  dmsg->type = type;
+  if (is_session_msg)
+    {
+      dmsg->seid = pfcp_session_msg_seid (p);
+      dmsg->seq_no = pfcp_session_msg_seq (p);
+    }
+  else
+    dmsg->seq_no = pfcp_node_msg_seq (p);
 
-  return encode_group (&msg_specs[type], grp, vec);
+  msg_len = pfcp_msg_length (p);
+  if (msg_len + 4 > len)
+    return PFCP_CAUSE_INVALID_LENGTH;
+
+  if (!is_session_msg != !pfcp_msg_s_flag (p))
```

Comment:
A bit of defensive programming for this "logical XOR". Right now both `is_session_msg` and `pfcp_msg_s_flag()` are guaranteed to be either `0` or `1`, but if this changes for some reason in future (yes it probably shouldn't but who knows), this may turn into a bug that may not be exactly easy to catch on the first glance.

Reply:
If that is your concern, wouldn't it be better to change `is_session_msg` and the return of `pfcp_is_session_msg` to type `bool` ? `pfcp_msg_s_flag` could be hardened by forcing its value to bool by changing its definition to :

```
#define pfcp_msg_s_flag(V) (!! (((pfcp_header_t *)(V))->s_flag))
```

Reply:
Thanks for the hint, did so. I'm too used to C89 still ;(